Sources/SwiftGRPC/Core/ClientNetworkMonitor.swift

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#if os(iOS)
+import CoreTelephony
+import Dispatch
+import SystemConfiguration
+
+/// This class may be used to monitor changes on the device that can cause gRPC to silently disconnect (making
+/// it seem like active calls/connections are hanging), then manually shut down / restart gRPC channels as
+/// needed. The root cause of these problems is that the backing gRPC-Core doesn't get the optimizations
+/// made by iOS' networking stack when changes occur on the device such as switching from wifi to cellular,
+/// enabling/disabling airplane mode, etc.
+/// Read more: https://github.com/grpc/grpc-swift/tree/master/README.md#known-issues
+/// Original issue: https://github.com/grpc/grpc-swift/issues/337
+open class ClientNetworkMonitor {
+    private let queue: DispatchQueue
+    private let callback: (State) -> Void
+    private var reachability: SCNetworkReachability?
+    private var appWasInForeground: Bool?
+
+    /// Instance of network info being used for obtaining cellular technology names.
+    public private(set) lazy var cellularInfo = CTTelephonyNetworkInfo()
+    /// Whether the network is currently reachable. Backed by `SCNetworkReachability`.
+    public private(set) var isReachable: Bool?
```

Comment:
Okay, now I see that this is also used to call the callback right after initialization, when we get initialized with the first state. I'm not a fan of optional bools, but I guess it works here.

Reply:
Right, these are only optional to avoid notifying when the initial state is set asynchronously. Not a huge fan of the optionals, but I thought this would yield the best behavior 🤷‍♂️